src/services/jsTyping.ts

```diff
@@ -0,0 +1,293 @@
+﻿// Copyright (c) Microsoft. All rights reserved. Licensed under the Apache License, Version 2.0.
+// See LICENSE.txt in the project root for complete license information.
+
+/// <reference path='services.ts' />
+
+/* @internal */
+namespace ts.JsTyping {
+
+    interface TypingResolutionHost {
+        directoryExists: (path: string) => boolean;
+        fileExists: (fileName: string) => boolean;
+        readFile: (path: string, encoding?: string) => string;
+        readDirectory: (path: string, extension?: string, exclude?: string[], depth?: number) => string[];
+    };
+
+    // A map of loose file names to library names
+    // that we are confident require typings
+    let safeList: Map<string>;
+    const notFoundTypingNames: string[] = [];
+
+    function tryParseJson(jsonPath: string, host: TypingResolutionHost): any {
+        if (host.fileExists(jsonPath)) {
+            try {
+                const contents = removeComments(host.readFile(jsonPath));
+                return JSON.parse(contents);
+            }
+            catch (e) { }
+        }
+        return undefined;
+    }
+
+    function isTypingEnabled(options: TypingOptions): boolean {
+        if (options) {
+            if (options.enableAutoDiscovery ||

```

Comment:
The `include` and `exclude` arrays need to be checked as well before skipping the typing discovery. Because the typings in `include` list may or may not be cached, and it is not correct to just return a empty array here. I would keep this function.